litentry/core/data-providers/src/twitter_official.rs

```diff
@@ -124,14 +126,65 @@ impl TwitterOfficialClient {
 		&mut self,
 		user: Vec<u8>,
 		original_tweet_id: Vec<u8>,
-	) -> Result<Tweets, Error> {
+	) -> Result<Tweet, Error> {
 		let original_tweet_id = vec_to_string(original_tweet_id)?;
 		let user = vec_to_string(user)?;
 		let query_value = format!("from: {} retweets_of_tweet_id: {}", user, original_tweet_id);
 		let query: Vec<(&str, &str)> =
 			vec![("query", query_value.as_str()), ("expansions", "author_id")];
-		self.client
-			.get_with::<String, Tweets>("/2/tweets/search/recent".to_string(), query.as_slice())
-			.map_err(|e| Error::RequestError(format!("{:?}", e)))
+		let resp = self
+			.client
+			.get_with::<String, TwitterAPIV2Response<Vec<Tweet>>>(
+				"/2/tweets/search/recent".to_string(),
+				query.as_slice(),
+			)
+			.map_err(|e| Error::RequestError(format!("{:?}", e)))?;
+		let tweets = resp.data.ok_or_else(|| Error::RequestError("tweet not found".to_string()))?;
+		if !tweets.is_empty() {
+			Ok(tweets.get(0).unwrap().clone())
+		} else {
+			Err(Error::RequestError("tweet not found".to_string()))
+		}
+	}
+
+	/// rate limit: 300/15min(per App) 900/15min(per User)
+	pub fn query_user(&mut self, user: Vec<u8>) -> Result<TwitterUser, Error> {
+		let user = vec_to_string(user)?;
+		let query = vec![("user.fields", "public_metrics")];
+		let resp = self
+			.client
+			.get_with::<String, TwitterAPIV2Response<TwitterUser>>(
+				format!("/2/users/{}", user),
+				query.as_slice(),
+			)
+			.map_err(|e| Error::RequestError(format!("{:?}", e)))?;
+		let user = resp.data.ok_or_else(|| Error::RequestError("user not found".to_string()))?;
+		Ok(user)
+	}
+}
+
+#[cfg(test)]
+mod tests {
+	use crate::twitter_official::TwitterOfficialClient;
+
+	#[test]
+	fn query_retweet_work() {
+		std::env::set_var("TWITTER_AUTHORIZATION_TOKEN", "Bearer ");
+		let mut client = TwitterOfficialClient::new();
+		let result = client.query_retweet(
+			"2934243054".as_bytes().to_vec(),
+			"1584490517391626240".as_bytes().to_vec(),
+		);
+		println!("query_retweet_work {:?}", result);
+		// assert!(result.is_ok(), "error: {:?}", result);
```

Comment:
Why not test the result? Is the twitter API unstable?

Reply:
Because we need to be set the `TWITTER_AUTHORIZATION_TOKEN` in env, and I don't want to leak the token in the code (or set it in the github secret?). I think we can get the unit tests to pass after the mockserver is ready.